prometheus-to-sd/config/dynamic_source.go

```diff
@@ -72,32 +95,62 @@ func createOptionsForPodSelection(nodeName string, sources map[string]url.URL) v
 		nameList += key + ","
 	}
 	labelSelector := fmt.Sprintf("%s in (%s)", nameLabel, nameList)
-	return v1.ListOptions{
+	listOptions := v1.ListOptions{
 		FieldSelector: fmt.Sprintf("spec.nodeName=%s", nodeName),
 		LabelSelector: labelSelector,
 	}
+	glog.Infof("Filters for dynamic components: %+v", listOptions)
+	return listOptions
 }
 
-func getConfigsFromPods(pods []core.Pod, sources map[string]url.URL) []SourceConfig {
-	var sourceConfigs []SourceConfig
-	for _, pod := range pods {
-		componentName := pod.Labels[nameLabel]
-		source, _ := sources[componentName]
-		podConfig := PodConfig{
-			PodId:       pod.Name,
-			NamespaceId: pod.Namespace,
-		}
-		sourceConfig, err := mapToSourceConfig(componentName, source, pod.Status.PodIP, podConfig)
-		if err != nil {
-			glog.Warning("could not create source config for pod %s: %v", pod.Name, err)
-		}
-		sourceConfigs = append(sourceConfigs, *sourceConfig)
+func getConfigFromPod(pod core.Pod, sources map[string]url.URL) (*SourceConfig, error) {
+	componentName := pod.Labels[nameLabel]
+	source, _ := sources[componentName]
+	podConfig := PodConfig{
+		PodId:       pod.Name,
+		NamespaceId: pod.Namespace,
 	}
-	return sourceConfigs
+	return mapToSourceConfig(componentName, source, pod.Status.PodIP, podConfig)
 }
 
 func mapToSourceConfig(componentName string, url url.URL, ip string, podConfig PodConfig) (*SourceConfig, error) {
 	port := url.Port()
 	whitelisted := url.Query().Get("whitelisted")
 	return newSourceConfig(componentName, ip, port, url.Path, whitelisted, podConfig)
 }
+
+func getSourceConfigChangeChannel(watchInterface watch.Interface, sources map[string]url.URL) <-chan SourceConfigChange {
+	sourceConfigChannel := make(chan SourceConfigChange)
+	eventChannel := watchInterface.ResultChan()
+	go func() {
+		for {
+			sourceConfigChannel <- mapPodWatchToSourceConfigChange(<-eventChannel, sources)
+		}
+	}()
+	return sourceConfigChannel
+}
+
+func mapPodWatchToSourceConfigChange(event watch.Event, sources map[string]url.URL) SourceConfigChange {
+	var sourceConfigChange SourceConfigChange
+	sourceConfigChange.EventObject = fmt.Sprintf("%v", event.Object)
+	if event.Type == watch.Error {
+		sourceConfigChange.ChangeType = Error
+		sourceConfigChange.Err = errors.New("got error when watching")
+		return sourceConfigChange
+	}
+	pod, ok := event.Object.(*core.Pod)
+	if !ok {
+		sourceConfigChange.ChangeType = Error
+		sourceConfigChange.Err = errors.New("not a pod object")
+		return sourceConfigChange
+	}
+	sourceConfig, err := getConfigFromPod(*pod, sources)
+	if err != nil {
+		sourceConfigChange.ChangeType = Error
+		sourceConfigChange.Err = err
```

Comment:
Do you need to create error events and send them instead of handling them locally?

Reply:
no, we can handle them locally too